docs/tutorials/Containers.md

```diff
@@ -178,27 +195,48 @@ To run Singularity on a job scheduler, the singularity command needs to be passe
 For example, in SLURM, we can use the normal SLURM configuration as explained in the [SLURM documentation](../backends/SLURM), however we'll add a `submit-docker` block to execute when a task is tagged with a docker container. 
 
 When constructing this block, there are a few things to keep in mind:
-- Make sure Singularity is loaded (and in our path), for example you can call `module load Singularity/3.0.1`.
-- We should treat worker nodes as if they do not have stable access to the internet or build access, so we will pull the container before execution of the task.
-- It's a good idea to ask Singularity to build the image into the execution directory of the task as an artifact, and to save rendering time on the worker node.
-- We'll use the docker name (potentially including the docker hash) to generate a compliant filename that can be looked up later.
-- If the container exists, there is no need to rebuild the container.
+- Make sure Singularity is loaded (and in our path), for example you can 
+  call `module load Singularity/3.0.1` if module is installed on your system.
+- We should treat worker nodes as if they do not have stable access to the 
+  internet or build access, so we will pull the container before execution of 
+  the task.
+- It's a good idea to use a Singularity cache so that same images should only
+  have to be pulled once. Make sure you set the `SINGULARITY_CACHEDIR` 
+  environment variable to a location on the filesystem that is reachable by the
+  worker nodes!
+- If we are using a cache we need to ensure that submit processes started by
+  Cromwell do not pull to the same cache at the same time. This may corrupt the
+  cache. We can prevent this by implementing a filelock with `flock` and 
+  pulling the image before the job is submitted. The flock and pull command 
+  needs to be placed *before* the submit command so all pull commands are 
+  executed on the same node. This is necessary for the filelock to work.
+- As mentioned above the `--containall` flag is **important** for 
```

Comment:
I'm not sure I totally agree with this. If you can't write files, that's a problem. containall becomes important when you have issues with path conflicts, etc.

Reply:
Containers implicitly loading the user's environment and home directory which may affect settings in certain tools is  a big problem. For reproducibility I only want settings that are explicitly given on the command line or otherwise via the WDL task to take effect.
Otherwise different users may have different results, meaning the workflow is not reproducable. In scientific workloads that is *completely unacceptable*. The use `--containall` recommendation can not be strong enough.
If you need functionality that is disabled by `--containall` you can always explicitly request this functionality. This is better in that you make your requirements explicit! Implicit requirements make it very hard to reproduce workflows.

Reply:
Cant argue with that. :) Why not mention —cleanenv too then?

Reply:
`--containall` implies `--cleanenv`. As well as `--contain` (no HOME mounting), `--ipc` (run in a new IPC namespace) and `--pid` (run in a new PID namespace). 

Technically using `--cleanenv --contain` also works. Except that some python tools (as @illusional mentioned) crash. This is probably (I am not certain) due to these python tools using the multiprocessing module which requires a separate IPC namespace. So `--ipc` and `--pid` should also be added to be sure.

Using `--containall` might also be translated as "Do not implicitly inherited anything from the host except the bare necessities. If I need more I will tell you so explicitly". I think that is good default behavior for containers to observe.

Reply:
okay let's briefly mention that the env is cleaned too - even the official [docs](https://sylabs.io/guides/3.1/user-guide/bind_paths_and_mounts.html) don't make that entirely clear.

Reply:
Done! I added a sentence that mentions this happening.